core/src/test/scala/unit/kafka/network/SocketServerTest.scala

```diff
@@ -337,11 +344,34 @@ class SocketServerTest {
       externalReadyFuture.complete(null)
       TestUtils.waitUntilTrue(() => listenerStarted(externalListener), "External listener not started")
     } finally {
+      externalReadyFuture.complete(null)
       executor.shutdownNow()
       shutdownServerAndMetrics(testableServer)
     }
   }
 
+  @Test
+  def testStagedListenerShutdownWhenConnectionQueueIsFull(): Unit = {
+    val testProps = new Properties
+    testProps ++= props
+    testProps.put("listeners", "EXTERNAL://localhost:0,INTERNAL://localhost:0,CONTROLLER://localhost:0")
+    testProps.put("listener.security.protocol.map", "EXTERNAL:PLAINTEXT,INTERNAL:PLAINTEXT,CONTROLLER:PLAINTEXT")
+    testProps.put("control.plane.listener.name", "CONTROLLER")
+    testProps.put("inter.broker.listener.name", "INTERNAL")
+    val config = KafkaConfig.fromProps(testProps)
+    val connectionQueueSize = 1
+    val testableServer = new TestableSocketServer(config, connectionQueueSize)
+    testableServer.startup(startProcessingRequests = false)
+
+    connect(testableServer, new ListenerName("EXTERNAL"), localAddr = InetAddress.getLocalHost)
+    connect(testableServer, new ListenerName("EXTERNAL"), localAddr = InetAddress.getLocalHost)
+
+    // Wait to let the acceptor accepts the connections
+    Thread.sleep(100)
```

Comment:
Can we replace sleep with some condition?

Reply:
I have reworked this test. It still consistently fails without this patch and it does not have the sleep any more.